File: day04/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"regexp"
	"sort"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	count1 := 0
	count2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		a, b, err := parseLine(scanner.Text())
		if err != nil {
			log.Fatalln(err)
		}
		if isFullyContained(a, b) {
			count1++
		} else if isOverlap(a, b) {
			count2++
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", count1)
	fmt.Println("Part 2:", count1+count2)
}

var (
	lineRegex = regexp.MustCompile(`^([0-9]+)-([0-9]+),([0-9]+)-([0-9]+)$`)
)

func parseLine(line string) (Pair, Pair, error) {
	match := lineRegex.FindStringSubmatch(line)
	if len(match) == 0 {
		return Pair{}, Pair{}, errors.New("Invalid line")
	}
	a, err := NewPairStr(match[1], match[2])
	if err != nil {
		return Pair{}, Pair{}, err
	}
	b, err := NewPairStr(match[3], match[4])
	if err != nil {
		return Pair{}, Pair{}, err
	}
	return a, b, nil
}

type (
	Pair struct {
		x, y int
	}
)

func NewPair(x, y int) Pair {
	if x < y {
		return Pair{
			x: x,
			y: y,
		}
	}
	return Pair{
		x: y,
		y: x,
	}
}

func NewPairStr(x, y string) (Pair, error) {
	i, err := strconv.Atoi(x)
	if err != nil {
		return Pair{}, err
	}
	j, err := strconv.Atoi(y)
	if err != nil {
		return Pair{}, err
	}
	return NewPair(i, j), nil
}

func isFullyContained(a, b Pair) bool {
	return isInclusive(a, b) || isInclusive(b, a)
}

func isInclusive(a, b Pair) bool {
	return b.x >= a.x && b.y <= a.y
}

func isOverlap(a, b Pair) bool {
	intervals := []Pair{
		{x: a.x, y: 0},
		{x: a.y + 1, y: 1},
		{x: b.x, y: 0},
		{x: b.y + 1, y: 1},
	}
	sort.Slice(intervals, func(i, j int) bool {
		a := intervals[i]
		b := intervals[j]
		if a.x != b.x {
			return a.x < b.x
		}
		if a.y != b.y {
			return a.y == 1
		}
		return false
	})

	count := 0

	for _, i := range intervals {
		if i.y == 0 {
			count++
			if count > 1 {
				return true
			}
		} else {
			count--
		}
	}

	return false
}

File: day04/src/main.rs
use lazy_static::lazy_static;
use regex::Regex;
use std::cmp::Ordering;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BoxResult<T> = Result<T, Box<dyn std::error::Error>>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let (count1, count2) = {
        let mut count1 = 0;
        let mut count2 = 0;
        for line in reader.lines() {
            let (a, b) = parse_line(&line?)?;
            if is_fully_contained(&a, &b) {
                count1 += 1;
            } else if is_overlap(&a, &b) {
                count2 += 1;
            }
        }
        (count1, count2)
    };
    println!("Part 1: {}\nPart 2: {}", count1, count1 + count2);
    Ok(())
}

fn parse_line(line: &str) -> BoxResult<(Pair, Pair)> {
    lazy_static! {
        static ref RE: Regex = Regex::new(r"^([0-9]+)-([0-9]+),([0-9]+)-([0-9]+)$").unwrap();
    }
    let captures = RE.captures(line).ok_or("Invalid line")?;
    Ok((
        Pair::from_str(
            captures.get(1).ok_or("Invalid line")?.as_str(),
            captures.get(2).ok_or("Invalid line")?.as_str(),
        )?,
        Pair::from_str(
            captures.get(3).ok_or("Invalid line")?.as_str(),
            captures.get(4).ok_or("Invalid line")?.as_str(),
        )?,
    ))
}

struct Pair(i32, i32);

impl Pair {
    fn new(x: i32, y: i32) -> Self {
        if x < y {
            Self(x, y)
        } else {
            Self(y, x)
        }
    }

    fn from_str(x: &str, y: &str) -> BoxResult<Self> {
        Ok(Self::new(x.parse::<i32>()?, y.parse::<i32>()?))
    }
}

fn is_fully_contained(a: &Pair, b: &Pair) -> bool {
    is_inclusive(a, b) || is_inclusive(b, a)
}

fn is_inclusive(a: &Pair, b: &Pair) -> bool {
    b.0 >= a.0 && b.1 <= a.1
}

fn is_overlap(a: &Pair, b: &Pair) -> bool {
    let mut intervals = vec![
        Pair(a.0, 0),
        Pair(a.1 + 1, 1),
        Pair(b.0, 0),
        Pair(b.1 + 1, 1),
    ];
    intervals.sort_unstable_by(|a, b| {
        if a.0 < b.0 {
            Ordering::Less
        } else if a.0 > b.0 {
            Ordering::Greater
        } else {
            if a.1 != b.1 {
                if a.1 == 1 {
                    Ordering::Less
                } else {
                    Ordering::Greater
                }
            } else {
                Ordering::Equal
            }
        }
    });

    let mut count = 0;
    for Pair(_, y) in intervals {
        if y == 0 {
            count += 1;
            if count > 1 {
                return true;
            }
        } else {
            count -= 1;
        }
    }
    false
}
